### ISSUE TITLE
The Road to 1.0

### ISSUE THREAD

**Issue by tomdale** [ember-fastboot/ember-cli-fastboot#98]
This is a quest issue tracking the features and bug fixes needed to consider FastBoot ready for mass consumption. Expect this list to grow over time. All discussion and details should be broken out into separate issues/PRs; this issue is just for coordinating and tracking work.
### Release Plan
- [x] [Expose FastBoot service](https://github.com/tildeio/ember-cli-fastboot/issues/79) (https://github.com/tildeio/ember-cli-fastboot/pull/121)
  - [x] Cookies
  - [x] Headers
- [x] [Support meta tags (Open Graph/Twitter cards)](https://github.com/tildeio/ember-cli-fastboot/issues/77)
- [x] ~~[Dockerized FastBoot server](https://github.com/tildeio/ember-cli-fastboot/issues/99)~~
- [x] [Ability to manually disable addons in FastBoot build](https://github.com/tildeio/ember-cli-fastboot/issues/57)
- [x] [Support query params](https://github.com/tildeio/ember-cli-fastboot/issues/53)
- [x] [Respect `outputPaths`](https://github.com/tildeio/ember-cli-fastboot/issues/36)
- [x] Support multi-stage (browser → FastBoot) build with asset rewriting, so FastBoot index.html points at browser build, not FastBoot build
- [ ] Documentation
  - [x] Conceptual
  - [x] Caveats/Known Limitations
  - [x] How to Deploy
  - [ ] ~~Production Debugging, Monitoring & Best Practices~~
  - [x] List of compatible ember-cli-deploy plugins
  - [x] [Addon author guide](https://github.com/tildeio/ember-cli-fastboot/issues/105)
  - [ ] Upgrade guide for app authors
- [ ] Fix issues with `watch` or revert feature for 1.0 release
- [ ] Website Updates
  - [ ] Update references: ember-fastboot-server -> fastboot-app-server
  - [ ] Component hooks that get invoked
  - [ ] ember-cli-head
  - [ ] Document `request` and `response` properties on the FastBoot service
  - [ ] Document `deferRendering` API
  - [ ] Document the shoebox
- [ ] Verify Node 0.12 compatibility
  - [ ] `fastboot`
  - [ ] `fastboot-express-middleware`
  - [ ] `fastboot-app-server`
  - [ ] `ember-cli-fastboot`
- [ ] Cut 1.0 releases
  - [ ] `fastboot`
  - [ ] `fastboot-express-middleware`
  - [ ] `fastboot-app-server`
  - [ ] `ember-cli-fastboot`
### Nice to Have
- [x] [Application control over HTTP status codes](https://github.com/tildeio/ember-cli-fastboot/issues/69)
- [x] Sample application
- [x] Tutorial

Please chime in with anything I've missed.

**Comment by danmcclain** [comment]
What about appending requested host to absolute paths in ajax requests

**Comment by tomdale** [comment]
@danmcclain Added https://github.com/tildeio/ember-cli-fastboot/issues/101 to track that.

**Comment by taveras** [comment]
How about being able to control over the `/robots.txt` response served from fastboot?

**Comment by tomdale** [comment]
@taveras It's just an Express middleware, so you can add your own handling for arbitrary endpoints. We should try not to let FastBoot turn into a generic HTTP server.

**Comment by tomdale** [comment]
Per [this week's meeting](https://github.com/ember-fastboot/meeting-notes/blob/master/2016-02-03.md):

- @ronco is working on adding support for `head.hbs` to render `<head>` elements
- @rwjblue is working on a change to Ember CLI so that config is written to a JSON file
- @cibernox is updating his addon to work in FastBoot and collecting a list of pain points for dual testing an addon in both scenarios
- @tomdale will work on exposing the Express request as `FastBoot.request`
- @RuslanZavacky is investigating getting good performance out of the Googlebot and is working on a PR that allows disabling `<script>` tags from FastBoot's HTML page.

**Comment by jpadilla** [comment]
I also started working on #99 :smiley:

**Comment by tomdale** [comment]
If you're interested in helping out, join us in #fastboot in the Ember community Slack.

**Comment by sebyx07** [comment]
it would also be nice at first to have cache headers in the response
```shell
ember fastboot --cache-headers 15000
```
before we have better access to headers

**Comment by tomdale** [comment]
@sebyx07 It's just an Express middleware, you can add whatever headers you want.

**Comment by MiguelMadero** [comment]
@tomdale what about "hydrating" or making the fastboot app into a normal emberApp after everything is loaded?

**Comment by danmcclain** [comment]
@MiguelMadero I think that hydrating is blocked until Glimmer 2.0 lands, I'll let someone else confirm

**Comment by tomdale** [comment]
@MiguelMadero @danmcclain Confirmed

**Comment by MiguelMadero** [comment]
Thanks for clarifying.

**Comment by danmcclain** [comment]
>Support multi-stage (browser → FastBoot) build with asset rewriting, so FastBoot index.html points at browser build, not FastBoot build

@tomdale Your latest PR achieves this, right?

**Comment by tomdale** [comment]
@danmcclain Confirm, good catch!

**Comment by coladarci** [comment]
Wow - when this is finished, Ember will really be able to DoItAll™️!
Realllllly want to start a new project w/ fastboot but am scared that w/out Hydration landing _relatively_ soon, it won't be the right decision and we might be left in a tough spot. Any ballpark ideas of when hydration will be added? I looked for a Glimmer 2.0 roadmap and couldn't find one.. Thanks for all the work on this - really a game-changer for Ember.

**Comment by hhff** [comment]
@coladarci - I think it's a safe bet - while fastboot doesn't technically 'rehydrate', it does 'redraw' the application once the browser JS payload hits.  It kinda clicks into place.  Works really well for the most part :+1:

**Comment by coladarci** [comment]
Yeah I noticed that, which is great. That said, that means for ~1.5s the actions on the page don't do anything, which is a bit of a blocker for my application :) I'd love to know where to follow the Glimmer 2.0 progress so I can get a rough idea of the blockers and how actively they are being addressed. We'll probably use ember for the new project w/out fastboot but checking it along the way to make sure there are no surprises, then hope we can flip it on before the launch.. I was thinking about a hybrid model where I remove a body class w/ the booted ember app so I can sort of gray out the body until the app is fully loaded; will give the end user immediate feedback and excitement but not render a bunch of busted actions for an awkward few seconds.

**Comment by tomdale** [comment]
@coladarci Yeah, one thing I'd like to do is automatically add an `ember-fastboot` class name to the `<body>` that gets removed once the app boots up. That would allow people to easily style interactive elements as disabled until the JavaScript has finished booting.

**Comment by Cryrivers** [comment]
@tomdale So 'hydrating' feature will come with Fastboot 1.0 and Ember 2.7?

**Comment by tomdale** [comment]
@Cryrivers I suspect it will not be in by FastBoot 1.0, but it's largely gated on Glimmer 2 work.

**Comment by oligriffiths** [comment]
@tomdale Am I right in thinking, according to the checklist above, that fastboot is essentially code ready, but needs the associated docs, website, watch, etc, sorting out?

**Comment by tomdale** [comment]
@oligriffiths Yes, we are very close. The last thing gating a 1.0 release is some performance improvements to the build process. After a lot of core team discussion, people were hesitant about the almost-doubling of build times that installing FastBoot adds to the boot process.

@stefanpenner had some great ideas for how we can avoid the extra cost, and I'm hopeful we can get those implemented soon. They will also clean up some of the monkeypatches to Ember CLI internals we have to do at the moment, and should address other usability concerns like #246.

**Comment by stefanpenner** [comment]
> @stefanpenner had some great ideas for how we can avoid the extra cost, and I'm hopeful we can get those implemented soon. They will also clean up some of the monkeypatches to Ember CLI internals we have to do at the moment, and should address other usability concerns like #246.

related: https://github.com/ember-fastboot/ember-cli-fastboot/issues/264#issuecomment-252276046

**Comment by kratiahuja** [comment]
@tomdale Here is an [RFC](https://github.com/ember-cli/ember-cli/issues/6350) that Stef and I worked on for fixing the fastboot builds. I would prefer to hold up 1.0 release until this is fixed since this a fully backward incompatible change to current fastboot ergonomics. If we don't hold, we will need to jump to Fastboot 2.0 and the experience may not be good. 

I will be working 100% of my time on that RFC and some other fastboot OSS stuff so I can get this out soon enough.

**Comment by simonihmig** [comment]
Could we get an update on what is still blocking a 1.0 release?

I guess https://github.com/ember-fastboot/ember-cli-fastboot/pull/356 is the major blocker here, which is hopefully landing soon, anything else that is not listed in the above to do list?

From that list, I think several items have not been updated yet...

>  Website Updates

Those things seem to have been addressed already, at least in some form. Anything missing there? Can help if needed...

> Verify Node 0.12 compatibility

This can be dropped, right?

**Comment by kratiahuja** [comment]
there is one more item: fixing this issue https://github.com/ember-fastboot/ember-cli-fastboot/issues/264 . There is a solution to address this and @leitsubomi is going to look at implementing it.

**Comment by fotinakis** [comment]
I'm curious, is there a security evaluation/penetration testing planned for fastboot v1? Has anyone attempted to break it?

I'm very excited about fastboot, just as a user I'm concerned about understanding the security implications of moving from a purely static-file serving production environment (which is relatively easy to serve and secure) to one where user requests will interact with a full dynamic server environment with node + middleware, which of course comes with a completely different security footprint.

**Comment by shahkashani** [comment]
Any word on whether rehydration will make it into 1.0, now that Glimmer 2.0 is out?

**Comment by danmcclain** [comment]
https://github.com/ember-fastboot/ember-cli-fastboot/issues/360 supersedes this issue, closing this as it is no longer up to date.

Rehydration is not currently a 1.0 blocker, but that doesn't prevent it from landing during the 1.0 series

**Comment by citizenconn** [comment]
@danmcclain I haven't been able to find any information with how ember + fastboot deals with rehydration. Does the app just rerender from scratch in the background and replace the existing dom?

**Comment by danmcclain** [comment]
@citizenconn Rehydration has yet to be implemented, so it has to rerender the app from scratch (or some from the shoebox) then replace the dom
